Source/Core/Core/IOS/USB/Common.cpp

```diff
@@ -26,6 +31,115 @@ void TransferCommand::FillBuffer(const u8* src, const size_t size) const
+{
+  Memory::Write_U16(return_value, static_cast<u32>(packet_sizes_addr + packet_num * sizeof(u16)));
+}
+
+u16 Device::GetVid() const
+{
+  return GetDeviceDescriptor().idVendor;
+}
+
+u16 Device::GetPid() const
+{
+  return GetDeviceDescriptor().idProduct;
+}
+
+bool Device::HasClass(const u8 device_class) const
+{
+  if (GetDeviceDescriptor().bDeviceClass == device_class)
+    return true;
+  const auto interfaces = GetInterfaces(0);
+  return std::any_of(interfaces.begin(), interfaces.end(), [&](const auto& interface) {
+    return interface.bInterfaceClass == device_class;
+  });
+}
+
+static void CopyToBuffer(std::vector<u8>* buffer, const void* data, const size_t size)
+{
+  buffer->insert(buffer->end(), reinterpret_cast<const u8*>(data),
+                 reinterpret_cast<const u8*>(data) + size);
+  const size_t number_of_padding_bytes = Common::AlignUp(size, 4) - size;
+  buffer->insert(buffer->end(), number_of_padding_bytes, 0);
+}
+
+static void CopyDescriptorToBuffer(std::vector<u8>* buffer, DeviceDescriptor descriptor)
+{
+  descriptor.bcdUSB = Common::swap16(descriptor.bcdUSB);
+  descriptor.idVendor = Common::swap16(descriptor.idVendor);
+  descriptor.idProduct = Common::swap16(descriptor.idProduct);
+  descriptor.bcdDevice = Common::swap16(descriptor.bcdDevice);
+  CopyToBuffer(buffer, &descriptor, descriptor.bLength);
+}
+
+static void CopyDescriptorToBuffer(std::vector<u8>* buffer, ConfigDescriptor descriptor)
+{
+  descriptor.wTotalLength = Common::swap16(descriptor.wTotalLength);
+  CopyToBuffer(buffer, &descriptor, descriptor.bLength);
+}
+
+static void CopyDescriptorToBuffer(std::vector<u8>* buffer, InterfaceDescriptor descriptor)
+{
+  CopyToBuffer(buffer, &descriptor, descriptor.bLength);
+}
+
+static void CopyDescriptorToBuffer(std::vector<u8>* buffer, EndpointDescriptor descriptor)
+{
+  descriptor.wMaxPacketSize = Common::swap16(descriptor.wMaxPacketSize);
+  // IOS only copies 8 bytes from the endpoint descriptor, regardless of the actual length
+  CopyToBuffer(buffer, &descriptor, sizeof(descriptor));
+}
+
+std::vector<u8> Device::GetDescriptorsUSBV4() const
+{
+  return GetDescriptors([&](const auto& descriptor) { return true; });
+}
+
+std::vector<u8> Device::GetDescriptorsUSBV5(const u8 interface, const u8 alt_setting) const
+{
+  return GetDescriptors([&](const auto& descriptor) {
+    // The USBV5 interfaces present each interface as a different device,
+    // and the descriptors are filtered by alternate setting.
+    return descriptor.bInterfaceNumber == interface && descriptor.bAlternateSetting == alt_setting;
+  });
+}
+
+std::vector<u8>
+Device::GetDescriptors(std::function<bool(const InterfaceDescriptor&)> predicate) const
+{
+  std::vector<u8> buffer;
+
+  const auto device_descriptor = GetDeviceDescriptor();
+  CopyDescriptorToBuffer(&buffer, device_descriptor);
+
+  const auto configurations = GetConfigurations();
+  for (u8 c = 0; c < static_cast<u8>(configurations.size()); ++c)
```

Comment:
Why is `size_t` not used here? `u8` seems a little weird. It would make more sense to cast where it's needed and not make the loop variable itself a u8. Ditto for `i` below.